lib/puppet/provider/package/puppetserver_gem.rb

```diff
@@ -50,6 +58,63 @@ def self.gemlist(options)
     end
   end
 
+  # The puppetserver gem cli commands are particularly slow as they start a JVM.
+  # Instead, for the often called list command, use the rubygems library from the puppet ruby,
+  #   setting the GEM_HOME and GEM_PATH to the values in the puppetserver configuration file.
+  # The rubygems library does not have access to java platform gems, for example: json (1.8.3 java),
+  #   but java platform gems should not be managed, by design.
+
+  def self.gem_list_local_via_rubygems(gem_regex)
+    pe_puppetserver_conf_file = '/etc/puppetlabs/puppetserver/conf.d/pe-puppet-server.conf'
+    os_puppetserver_conf_file = '/etc/puppetlabs/puppetserver/puppetserver.conf'
+    puppetserver_gem_home = '/opt/puppetlabs/server/data/puppetserver/jruby-gems'
+    puppetserver_gem_path = ['/opt/puppetlabs/server/data/puppetserver/jruby-gems', '/opt/puppetlabs/server/data/puppetserver/vendored-jruby-gems']
+
+    gem_env = {}
+
+    if Facter.value(:pe_server_version)
+      puppetserver_conf_file = pe_puppetserver_conf_file
+    else
+      puppetserver_conf_file = os_puppetserver_conf_file
+    end
+
+    puppetserver_conf = Hocon.load(puppetserver_conf_file)
+
+    if puppetserver_conf.empty?
+      gem_env['GEM_HOME'] = puppetserver_gem_home
+      gem_env['GEM_PATH'] = puppetserver_gem_path
+      gem_env['GEM_SPEC_CACHE'] = '/dev/null'
+    else
+      gem_env['GEM_HOME'] = puppetserver_conf['jruby-puppet']['gem-home']
+      gem_env['GEM_PATH'] = puppetserver_conf['jruby-puppet']['gem-path'].join(':')
+      gem_env['GEM_SPEC_CACHE'] = '/dev/null'
+    end
+
+    Gem.paths = gem_env
+
+    sio_in, sio_out, sio_err = StringIO.new, StringIO.new, StringIO.new
+    strm_io = Gem::StreamUI.new(sio_in, sio_out, sio_err, false)
+
+    gem_list_cmd = Gem::Commands::ListCommand.new
+    gem_list_cmd.ui = strm_io
+    gem_list_cmd.options[:args] = [gem_regex] if gem_regex
+    gem_list_cmd.execute
+
+    # Remove default gems from the list, as the are the default gems from the puppet ruby:
+    #   /opt/puppetlabs/puppet/lib/ruby/gems/x.y.z/specifications/default
+    # There is no method exclude default gems from the local gem list, for example: psych (default: 2.2.2),
+    #   but default gems should not be managed, by design.
+
+    gem_list = sio_out.string.lines.reject { |gem| gem =~ / \(default\: / }
+    gem_list.join("\n")
+
+    ensure
```

Comment:
Tiny style nit: one level of indentation should be removed here so that `ensure` lines up with `def` and `end`.

Reply:
Mos def! :)